src/hotspot/share/runtime/lightweightSynchronizer.cpp

```diff
@@ -628,16 +633,33 @@ void LightweightSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* cur
     return;
   }
 
-  const int spins = OMSpins;
-  const int yields = OMYields;
+  // Will spin with exponential backoff with an accumulative O(2^spin_limit) spins.
+  const int log_spin_limit = os::is_MP() ? OMSpins : 1;
+  const int log_min_safepoint_check_interval = 10;
 
   while (true) {
-
-    SpinYield fast_lock_spin_yield(spins, yields);
     // Fast-locking does not use the 'lock' argument.
-    markWord mark = obj()->mark_acquire();
+    markWord mark = obj()->mark();
     const bool try_spin = !first_time || !mark.has_monitor();
-    for (int attempts = spins + yields; try_spin && attempts > 0; attempts--) {
+    // Always attempt to lock once even when safepoint synchronizing.
+    bool should_process = false;
+    for (int i = 0; try_spin && !should_process && i < log_spin_limit; i++) {
+      // Spin with exponential backoff.
+      const int total_spin_count = 1 << i;
+      const int inner_spin_count = MIN2(1 << log_min_safepoint_check_interval, total_spin_count);
+      const int outer_spin_count = total_spin_count / inner_spin_count;
+      for (int outer = 0; outer < outer_spin_count; outer++) {
+        should_process = SafepointMechanism::should_process(current);
+        if (should_process) {
+          // Stop spinning for safepoint.
+          break;
+        }
+        for (int inner = 1; inner < inner_spin_count; inner++) {
+          SpinPause();
+        }
+      }
+
+      mark = obj()->mark();
```

Comment:
Is this the spinning to delay creating an ObjectMonitor for this lock as long as possible?  Or if deflation is observed?  This seems like it'd slow down the contended case, like xalan where we want a lot of the threads to quickly park.  Or is this spinning because deflation is observed (like the comment says above)?

Can you make 647-660 a function above this so it doesn't distract from the logic of the rest of enter so much?

Reply:
> Is this the spinning to delay creating an ObjectMonitor for this lock as long as possible? Or if deflation is observed? This seems like it'd slow down the contended case, like xalan where we want a lot of the threads to quickly park. Or is this spinning because deflation is observed (like the comment says above)?

There are two spinning loops here. The outer which uses the SpinYield which is only to do with deflation. And the inner when the object is fast_lock. But the first_time variable has a bad name. And with the new exponential backoff changes the spinning should be interrupted if inflation is observed.


> Can you make 647-660 a function above this so it doesn't distract from the logic of the rest of enter so much?

Good suggestion. Becomes much cleared. 

Extracted the spinning logic, rename the variables and cleaned up the loop condition.